src/tasks/Tasks.ts

```diff
@@ -0,0 +1,961 @@
+    // Cancel the timer if it exists
+    this.schedulingTimer?.cancel();
+    // Cancel the scheduling loop
+    this.schedulingLoop.cancel(abortSchedulingLoopReason);
+    // Wait for the cancellation signal to resolve the promise
+    await this.schedulingLoop;
+    // Indicates that the loop is no longer running
+    this.schedulingLoop = null;
+    this.logger.info('Stopped Scheduling Loop');
+  }
+
+  protected async stopQueueing() {
+    if (this.queuingLoop == null) return;
+    this.logger.info('Stopping Queuing Loop');
+    // Cancel the scheduling loop
+    this.queuingLoop.cancel(abortQueuingLoopReason);
+    await this.queuingLoop;
+    this.queuingLoop = null;
+    this.logger.info('Stopped Queuing Loop');
+  }
+
+  /**
+   * Triggers the scheduler on a delayed basis
+   * If the delay is 0, the scheduler is triggered immediately
+   * The scheduling timer is a singleton that can be set by both
+   * `this.schedulingLoop` and `this.scheduleTask`
+   * This ensures that the timer is set to the earliest scheduled task
+   */
+  protected triggerScheduling(scheduleTime: number) {
+    if (this.schedulingTimer != null) {
+      if (scheduleTime >= this.schedulingTimer.scheduled!.getTime()) return;
+      this.schedulingTimer.cancel();
+    }
+    const now = Math.trunc(performance.timeOrigin + performance.now());
+    const delay = Math.max(scheduleTime - now, 0);
+    // On the first iteration of the scheduling loop
+    // the lock may not be acquired yet, and therefore releaser is not set
+    // in which case don't do anything, and the lock remains unlocked
+    if (delay === 0) {
+      this.schedulerLogger.debug(
+        `Setting scheduling loop iteration immediately (delay: ${delay} ms)`,
+      );
+      this.schedulingTimer = null;
+      if (this.schedulingLockReleaser != null) {
+        this.schedulingLockReleaser();
+      }
+    } else {
+      this.schedulerLogger.debug(
+        `Setting scheduling loop iteration for ${new Date(
+          scheduleTime,
+        ).toISOString()} (delay: ${delay} ms)`,
+      );
+      this.schedulingTimer = new Timer(() => {
+        this.schedulingTimer = null;
+        if (this.schedulingLockReleaser != null) {
+          this.schedulingLockReleaser();
+        }
+      }, delay);
+    }
+  }
+
+  /**
+   * Same idea as triggerScheduling
+   * But this time unlocking the queue to proceed
+   * If already unlocked, subsequent unlocking is idempotent
+   * The unlocking of the scheduling is delayed
+   * Whereas this unlocking is not
+   * Remember the queuing just keeps running until finished
+   */
+  protected triggerQueuing() {
+    if (this.taskPromises.size >= this.activeLimit) return;
+    // On the first iteration of the queuing loop
+    // the lock may not be acquired yet, and therefore releaser is not set
+    // in which case don't do anything, and the lock remains unlocked
+    if (this.queuingLockReleaser != null) {
+      this.queuingLockReleaser();
+    }
+  }
```

Comment:
`if (this.taskPromises.size >= this.activeLimit) return;` Shouldn't this be checking against `activePromises` for the concurrency limit? `taskPromises` tracks the promises wrapping the event system.

Reply:
Yes that should be `activePromises`. Forgot to change.